nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/util/SecureNiFiConfigUtil.java

```diff
@@ -0,0 +1,190 @@
+            TlsConfiguration tlsConfiguration = null;
+            cmdLogger.info("Generating Self-Signed Certificate: Expires on {}", LocalDate.now().plus(CERT_DURATION_DAYS, ChronoUnit.DAYS));
+            try {
+                String[] subjectAlternativeNames = getSubjectAlternativeNames(nifiProperties, cmdLogger);
+                tlsConfiguration = KeyStoreUtils.createTlsConfigAndNewKeystoreTruststore(StandardTlsConfiguration
+                        .fromNiFiProperties(nifiProperties), CERT_DURATION_DAYS, subjectAlternativeNames);
+            } catch (IOException | GeneralSecurityException e) {
+                cmdLogger.error("Self-Signed Certificate Generation Failed", e);
+                throw e;
+            }
+
+            // Move over the new stores from temp dir
+            Files.move(Paths.get(tlsConfiguration.getKeystorePath()), Paths.get(keystorePath),
+                    StandardCopyOption.REPLACE_EXISTING);
+            Files.move(Paths.get(tlsConfiguration.getTruststorePath()), Paths.get(truststorePath),
+                    StandardCopyOption.REPLACE_EXISTING);
+
+            updateProperties(propertiesFile, tlsConfiguration);
+
+            cmdLogger.debug("Successfully generated {} and {}.", keystorePath, truststorePath);
+        } else if (!keystoreExists && truststoreExists) {
+            cmdLogger.warn("Truststore file {} already exists.  Apache NiFi will not generate keystore and truststore separately.",
+                    truststorePath);
+        } else if (keystoreExists && !truststoreExists) {
+            cmdLogger.warn("Keystore file {} already exists.  Apache NiFi will not generate keystore and truststore separately.",
+                    keystorePath);
+        }
+    }
+
+    /**
+     * Attempts to add some reasonable guesses at desired SAN values that can be added to the generated
+     * certificate.
+     * @param nifiProperties The nifi.properties
+     * @return A Pair with IP SANs on the left and DNS SANs on the right
+     */
+    private static String[] getSubjectAlternativeNames(Properties nifiProperties, Logger cmdLogger) {
+        Set<String> dnsSubjectAlternativeNames = new HashSet<>();
+
+        try {
+            dnsSubjectAlternativeNames.add(InetAddress.getLocalHost().getHostName());
+        } catch (UnknownHostException e) {
+            cmdLogger.debug("Could not add localhost hostname as certificate SAN", e);
+        }
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.REMOTE_INPUT_HOST, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.WEB_HTTPS_HOST, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.WEB_PROXY_HOST, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.LOAD_BALANCE_ADDRESS, dnsSubjectAlternativeNames);
+
+        // Not necessary to add as a SAN
+        dnsSubjectAlternativeNames.remove(LOCALHOST_NAME);
```

Comment:
Is there a reason that this should be removed?  Is it because `localhost` is included as the Common Name?  It seems simpler to leave it in as a SAN, but I may be missing something.

Reply:
Because CertificateUtils.generateSelfSignedX509Certificate(...) already adds the CN as a SAN, so localhost will already be a SAN.